cli: gracefully handle SLS rendering errors

rjfd opened the pull request:
The deepsea-cli was not handling the case when some SLS file contained syntax errors, and an ugly python stack trace was being thrown.

This PR implements the handling of SLS syntax errors.

Video with demo: https://asciinema.org/a/ap5G06hRIxR2JO2x6YOp7hAY6

Signed-off-by: Ricardo Dias <rdias@suse.com>

smithfarm commented:
@rjfd This is a nit and it doesn't relate directly to this PR, but . . . I noticed that `check_root_privileges` exits immediately, and it's called on every invocation of deepsea, even ones that include the `--help` option. So, for example:

```
cephadm@ip-10-0-4-10:~> deepsea --help
Root privileges are required to run this tool
cephadm@ip-10-0-4-10:~> deepsea stage --help
Root privileges are required to run this tool
cephadm@ip-10-0-4-10:~> deepsea foobarbaz
Root privileges are required to run this tool
```

smithfarm commented:
**Test runs of interest**

These two runs include #620, #632, and #643 in addition to this PR:

* http://149.202.189.253:8081/ubuntu-2017-09-16_09:19:24-deepsea:basic-ses5---basic-openstack/
* http://149.202.189.253:8081/ubuntu-2017-09-16_10:15:31-deepsea:basic:rgw-ses5---basic-openstack/

Just this PR and #643:

* http://149.202.189.253:8081/ubuntu-2017-09-16_10:57:16-deepsea:basic:rgw-ses5---basic-openstack/

tserong commented on cli/stage_parser.py:
```diff
@@ -49,9 +94,18 @@ def render(file_name):
+                try:
+                    result = __caller__.cmd('slsutil.renderer', file_name)
+                except salt.exceptions.SaltException as ex:
+                    exception = StageRenderingException(file_name, ex.strerror)
```
Why not raise the exception here? Or does that cause trouble becuase stderr and stdout are redirected?

rjfd replied:
Right, if I raise the exception here, all the output made by the code that catches the exception is redirected.

tserong commented on cli/stage_parser.py:
```diff
@@ -35,6 +36,50 @@ class OrchestrationNotFound(Exception):
+        """
+        output = ""
+        if self.error_desc.find("SaltRenderError") == -1:
+            found_ex_line = True
```
I don't understand this piece. That reads to me as if it would set found_ex_line = True if "SaltRenderError" was *not* found? Then later, it checks if each line starts with "SaltRenderError" anyway?

rjfd replied:
The idea here is to remove the stack trace part of a `SaltRenderError` exception and keep the rest of the exception output, but keep the stack trace part of any other exception that might be caught.

I agree that the way this is written is confusing, I'll change this into a more readable piece of code.

tserong commented on cli/stage_executor.py:
```diff
@@ -71,12 +73,30 @@ def run_stage(stage_name, hide_state_steps, hide_dynamic_steps, simple_output):
+            PP.println(PP.cyan("    {}".format(ex.stage_file)))
+        PP.println()
+        PP.println(PP.bold("Error description:"))
+        PP.println(PP.red(ex.pretty_error_desc_str()))
```
The above block looks like it's the same as the block in start_stage() in monitor.py. Can these two be sanely put into a separate helper function?

rjfd replied:
Yes, you're right. I'll refactor this to have a single location with the prints for this kind of exception.

rjfd commented:
@smithfarm 

>  noticed that check_root_privileges exits immediately, and it's called on every invocation of deepsea, even ones that include the --help option.

Yes, need to fix that. Currently every code path checks if the user is "root", even when just using the --help option.
I also need to make the CLI to run the "stage parsing" phase as the "salt" user, otherwise the when runners are executed due to the rendering of sls files, some runner may create files, which only root can read or write.

rjfd commented:
@tserong addressed your comments, and pushed to the PR.
I moved the output code to the place where it should belong, which is in the `terminal_outputter.py` module.

rjfd commented:
@smithfarm care to run the teuthology suite to see if the latest changes didn't break anything?

smithfarm commented:
**running** http://149.202.189.253:8081/ubuntu-2017-09-21_12:04:29-deepsea:basic-ses5---basic-openstack/
